### PR TITLE
load balancer: added least request/random/round robin load balancer extension

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: load balancer: added least request/random/round robin load balancer extension
Additional Description:

Part of work of #20634. This PR make the  least request/random/round robin load balancer as an extension and could be configured by the `load_balancing_policy`.

One-page simple roadmap doc about the new extensions.
https://docs.google.com/document/d/1GtayHX8orSEChtdn_2stRB8BNY2LE0Y0FatDVGODNHA/edit?usp=sharing

Risk Level: Low. Not change current behavior.
Testing: wait.
Docs Changes: wait.
Release Notes: wait.
Platform Specific Features: wait.

### CODE REVIEW

**Comment by wbpcode** [comment]
/wait

Waiting for #23474 #23453 to be merged first.

**Comment by wbpcode** [comment]
cc @markdroth This is the first implementation for the load balancer extension. And it would be very quick to convert the other policies to the extension implementation.

In this implementation, I treat the `common_lb_config` a shared context of target cluster. Would this be acceptable for you?

**Comment by wbpcode** [comment]
Hi, @mattklein123 this is ready for a formal code review 😺

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. A few questions. Thank you.

/wait

File: source/common/upstream/load_balancer_impl.h
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h:624]: Sorry for there is no specified related doc for it for now. I will clarify it it this PR or the original issue.

These are some short replies to your questions.

> The old config is still supported, right?

Yes, forever, unless one day we change our minds and decide to cut out v4 API.

> So shouldn't the old config create a synthetic typed config to allow for migration and then should that config be marked deprecated?

It will but not for now. This PR just make the `load_balancing_policy` work for least request policy. But It couldn't be used to replace original LB enum `LEAST_REQUEST` for now. There still some common problems need to be solved, e.g. the recreation of LB instance, subset support.
So I don't think we can mark the old configuration as deprecated yet. And as you said, it's will cause a huge amount of WARNING for end users, So I am not sure could we dreprecate it finally.

> I guess i'm trying to understand how old and new are going to coexist? Is there some detail somewhere that I can read? Thank you.

Both old config and new config will worked and the new config will be taken as a priority.

**Comment by markdroth** [comment]
> In this implementation, I treat the `common_lb_config` a shared context of target cluster. Would this be acceptable for you?

I don't quite understand what that means.  Can you say more about what you mean by that?  Maybe provide an example of how `common_lb_config` would be used?

The main issue here is that the new extension-based LB policy API allows for hierarchical LB policies, where the top-level policy may have one or more child policies that it delegates to under certain circumstances.  For example, the `WrrLocality` LB policy will have a child policy for each locality, where the top-level policy will pick a locality and then delegate to a child policy for that locality, and the child policy will then pick the endpoint within that locality.  In that model, I don't think `common_lb_config` makes sense, because it isn't necessarily the case that all parameters will be the same for all policies in the tree.

I am not very familiar with every single field in `common_lb_config` or with their implementations in Envoy, but in principle, I think the fields in `common_lb_config` should fit into one of two categories:

1. Fields like `override_host_status`, which make sense only at the top level, since they effectively override the entire LB policy tree.  (The fact that this functionality happens to be implemented in the base LB policy class in Envoy is IMHO a bit ugly, since it doesn't make sense for anything but the top-level LB policy.  But ultimately, that's an implementation detail.)  I think these fields should be pulled out of `common_lb_config` and should instead be top-level fields in the `Cluster` message.
2. Fields like `use_hostname_for_hashing` and `hash_balance_factor` (currently nested in the `consistent_hashing_lb_config` field), which affect only hash-based LB policies.  I think these fields should be copied into the extension configs for the specific policies that use them.  (Note that we already did this for the `RingHash` policy.)

I think someone should audit all of the fields in `common_lb_config` and make sure that they all fit into one of the two above categories.  If there are any other categories, we need to identify them and figure out what to do about them.

I'd really like to get to a point where the existing `common_lb_config` field is not visible at all to extension-based LB policies in Envoy.

**Comment by wbpcode** [comment]
I am not sure to get it clear. If I am wrong, feel free to point it out.

IMO, the hierarchical LB policies are the implementation detail of one special top LB policy. And for the envoy, every cluster only have a extendable policy.

In my impl, the `common_lb_config` just be treated as part of cluster and could be accessed by any policy. **It could be anything that shared by the all policies.**

I agree the fields that only effect part of policies should be move to the extension-specific configuration. But the other field that could effect all these policies (top) should be ok to stay there.

**Comment by markdroth** [comment]
> IMO, the hierarchical LB policies are the implementation detail of one special top LB policy. And for the envoy, every cluster only have a extendable policy.

It's true that every cluster will have one top-level policy and that it's up to that policy to decide whether to delegate to other policies.  However, the `WrrLocality` policy is just one example of a policy that will delegate to other policies; it is not the only one that will ever exist.  I believe that hierarchical routing will become more useful over time.  We already use it heavily in gRPC.

> In my impl, the `common_lb_config` just be treated as part of cluster and could be accessed by any policy. **It could be anything that shared by the all policies.**

Can you give me an example of a field that you think would need to be (a) accessed by all of the policies and (b) will never need to be set differently for different policies in the tree?  I am not aware that any such a field exists.

For example, consider the consistent-hashing fields that I mentioned earlier.  If we wanted to write a hash-based locality-picking policy that would delegate to `RingHash` for endpoint-picking, we might want to set these values differently for the locality-picking policy than for the endpoint-picking policy -- or we might want to set them differently for different endpoint-picking policies underneath the locality-picking policy.

In general, I think all fields in `common_lb_config` should fit into one of the two categories I mentioned earlier.  I am open to hearing a counter-example if you can find one.

**Comment by mattklein123** [comment]
> In general, I think all fields in common_lb_config should fit into one of the two categories I mentioned earlier. I am open to hearing a counter-example if you can find one.

I agree with @markdroth that this is the type of audit that we need to do. I'm really hesitant to go down this path further until we figure out the final API mappings because otherwise we will just create more mess that needs to be cleaned up later?

**Comment by wbpcode** [comment]
> Can you give me an example of a field that you think would need to be (a) accessed by all of the policies and (b) will never need to be set differently for different policies in the tree? I am not aware that any such a field exists.

cc @markdroth 

Sorry, maybe I made some unclear descriptions. What I mean by `It could be anything that shared by all policies` is something that will worked when whatever policy is configured and will affect the finally load balancing result (the result from the users' perspective). 
For example, `override_host_status`, first category field as you mentioned. (Off course, this field will be accessed by all top policies for now, but it's just implementation detail and we could optimized it in the future. More important is, from user's perspective, it's something will affect the load balancing result under any policy.) 

**In fact, as I said before, I agree almost all your points except the moving for first category fields.**
Especially, considering the compatibility, we need keep the `common_lb_config` forever (because there will no v4 API).
This moving for first category fields just make thing more complex and has no much benefit.

**Comment by wbpcode** [comment]
cc @markdroth @mattklein123 
I agree we could do a complete audit for all field in the `common_lb_config` to classify them.

And 

- For the first category fields that works whatever policy is configured for the cluster (policy independent), just let them stay there. (cc @markdroth I think that's only area where we disagree?)
- For the second category fields that works under specific policy, copy it to the configuration of the policy extension.
- Make sure all fields in the `common_lb_config` still works for all exist policies' extensions (round robin, least request, etc.) for back-compatibility.

**Comment by wbpcode** [comment]
By the way, in fact, above discussion should have no big influence on this PR. This PR just wrapped the constructor of `least request lb` as a factory that could be used as policy extension.

We have a lot of things to finally achieve our final targe.
- First, we need make all other exist policies could be used as extension. (This PR just part of this work)
- Second, we may need clean the `common_lb_config` to move policy-dependent fields to extension configuration.
- Third, subset lb support.
- Ensure our extensions could be used to replace current implementations and may try to deprecate the old configuration.

**Comment by markdroth** [comment]
I don't think the concern here is actually directly about whether these fields stay in `CommonLbConfig`; I think it's more about how we handle `CommonLbConfig` semantically.

I agree with you that we need to retain backward compatibility.  The semantic that we discussed and implemented in #17400 was that `CommonLbConfig` would continue to be passed to the LB policy implementation when the LB policy is configured via the old-style enum field (i.e., we did not break any existing configurations), but it is not used at all when the LB policy is configured via the new extension mechanism.  I think the question we're facing now is how we support `CommonLbConfig` in the new extension mechanism, specifically for the fields that make sense only at the top level (category 1 above).

I think the core problem here is that an LB policy extension may be invoked either as a top-level policy or as a child policy, and the LB policy implementation should not know or care which one it is.  Parent policies will use the same API as the cluster implementation does to access the child policy, and it should not matter to a given LB policy implementation whether its parent is the cluster implementation or another LB policy.  Given that, I don't think it makes sense for `CommonLbConfig` to be passed to that API, because it is inherently non-sensical to use those fields anywhere but the top level.

In principle, here's how I'd like to see us proceed:
1. Move the top-level functionality out of the LB policy base class and into the cluster implementation, so that no LB policy implementation needs to look at the category 1 fields in `CommonLbConfig`.  (Disclaimer: I am not sufficiently familiar with the Envoy code to know whether or not this is actually feasible.)
2. Implement all LB policies as extensions, each with its own extension config.  The extension config for a given policy should directly contain any category 2 fields from `CommonLbConfig` that are relevant to that policy.
3. When we receive a `Cluster` resource that uses the old-style LB policy config (enum-based policy selection and built-in config messages nested in the `Cluster` resource), we convert the old-style config to the new extension config.  This will include copying category 2 fields from `CommonLbConfig` to the extension config.  We then instantiate the relevant extension with the converted extension config.
4. When we receive a `Cluster` resource with a new-style LB config, we directly instantiate the LB policy extension and pass it the provided extension config as-is.  We do not use `CommonLbConfig` for this.
5. At this point, `CommonLbConfig` should be completely unused by LB policy implementations, so we can remove it from the API, thus ensuring that it will never be used incorrectly.

If the above is not realistic, I'm open to alternative suggestions here, as long as they satisfy the key requirement that we have a structural and consistent way to ensure that (a) `CommonLbConfig` category 1 fields get used only at the top level and (b) category 2 fields never get used directly.  I don't think we can just rely on individual LB policy authors to get this right; I think it's going to be too hard to understand what the right behavior is, and we're going to wind up getting it wrong too often, which is going to cause a major headache for compatibility with other data planes like gRPC.

I agree in principle that from an API perspective, it doesn't matter that much whether these category 1 fields are directly in the `Cluster` message or remain in the `CommonLbConfig` sub-message.  I think the real question is having a structural way to ensure that `CommonLbConfig` does not get used incorrectly.  If we can do that, then I'm fine leaving those fields where they currently are.

**Comment by wbpcode** [comment]
>  I think the real question is having a structural way to ensure that CommonLbConfig does not get used incorrectly. If we can do that, then I'm fine leaving those fields where they currently are.

I was going to leave that up to the developers of policies, but you're right, maybe that's not appropriate.

We'll do an audit first to decide how to proceed, and I have a feeling it's going to be a lot of work.

**Comment by wbpcode** [comment]
cc @markdroth @mattklein123 

Some results of audit of `common_lb_config`.

- audit result of `common_lb_config`:  
	- `healthy_panic_threshold`: used for panic mode calculation. It affects all load balancing policies but still depends on implementation of load balancing policies.  
	- `zone_aware_lb_config`: For specific load balancing policies. Easy to be moving to related policies' configuration.  
	- `locality_weighted_lb_config`: ditto.  
	- `consistent_hashing_lb_config`: ditto.  
	- `update_merge_window`: used by the membership updating and policy could ignore this field.  
	- `ignore_new_hosts_until_first_hc`: ditto.  
	- `close_connections_on_host_set_change`: ditto.  
	- `override_host_status`: used for host overriding. It affects all (top) load balancing policies but still depends on implementation of load balancing policies. Maybe we can optimize the implementation to make sure this feature completely policy-independent. Then we can ignore it safely.  
- Some conclusions of the audit:  
	- There are four classes' fields actually:  
		- 1) field affects all load balancing policies, `healthy_panic_threshold`. But it still depends on policy implementation because may be the sub policies won't re-calculate panic mode.  
		- 2) field affects all (top) load balancing policies, `override_host_status`.  
		- 3) fields affect specific policies.  
		- 4) fields completely policy-independent and could be ignored in the policies implementation.  
	- In the perfect world, maybe we could optimize the implementation of upstream module and make the fields of class a and class b completely policy-independent. But in fact, it would a ton of work and hard to complete in a short time (especially for the class a).
	- We can make the fields of class c be handled correctly first.

**Comment by wbpcode** [comment]
cc @mattklein123 one-page simple doc about `how old and new are going to coexist?`. Google doc is used because it's easy to iterate it.

https://docs.google.com/document/d/1GtayHX8orSEChtdn_2stRB8BNY2LE0Y0FatDVGODNHA/edit?usp=sharing

**Comment by wbpcode** [comment]
/wait

waiting for most things to be clarified.

**Comment by wbpcode** [comment]
/wait for #23838 first. I will do some refactor after #23838 is merged.

**Comment by markdroth** [comment]
Thanks for doing that audit!

Is there a reason that we can't treat `healthy_panic_threadhold` as a policy-specific field (category c)?  It seems to me like that could be something implemented by specific policies as they pass control down through the tree, to decide which subset of hosts they pass to the child policy.  But I don't think it necessarily affects all LB policies, because there could be policies that don't implement panic mode.  (gRPC does not have the concept of panic mode at all.)

With regard to `override_host_status`, if we don't have a path toward pulling that out of the LB policy base class and into the cluster implementation, then how do we ensure that new LB policy extensions are not going to misuse `CommonLbConfig`?  I think we need some story around this.

**Comment by wbpcode** [comment]
> 

Almost every LB in the Envoy implemented this panic mode. And I don't think different value is necessary for the sub LBs.

(Although tell the truth, the panic mode (and failover) is too complex and flexible)

**Comment by wbpcode** [comment]
#23872 will make the `override_host_status` policy independent.

**Comment by markdroth** [comment]
It seems to me that we need to treat `healthy_panic_threadhold` as either category c or as category d.  Otherwise, we don't have a story here that does not require passing `CommonLbConfig` into the LB policy.

I'm certainly open to alternative suggestions.

**Comment by wbpcode** [comment]
> It seems to me that we need to treat `healthy_panic_threadhold` as either category c or as category d. Otherwise, we don't have a story here that does not require passing `CommonLbConfig` into the LB policy.
> 
> I'm certainly open to alternative suggestions.

category d would be great. But it would be a very big refactor to the envoy upstream module. 🤔 
Considering this value should never be different for top lb policies or sub lb policies, it should be OK to just let it there and do nothing for now and leave this to the end.

**Comment by RyanTheOptimist** [comment]
test/per_file_coverage.sh changes, LGTM

**Comment by wbpcode** [comment]
cc @markdroth for API change.
cc @mattklein123 for code change.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
At a high level makes sense. Can we get integration tests for the new extensions? Thank you.

/wait

**Comment by alyssawilk** [comment]
Oh this is awesome - I'd had this on my TODO List for Envoy size work.
Question - as part of this (probably in a follow up) would you be willing to have the load balancer code be factory created, so the code for each load balancer can be removed to the respective extension directory and out of the main build, or is that out of scope for you?

**Comment by wbpcode** [comment]
> Question - as part of this (probably in a follow up) would you be willing to have the load balancer code be factory created, so the code for each load balancer can be removed to the respective extension directory and out of the main build, or is that out of scope for you?

It's in my TODO list.

**Comment by wbpcode** [comment]
> At a high level makes sense. Can we get integration tests for the new extensions?

Get it.

File: changelogs/current.yaml
Comment (wbpcode) [reply changelogs/current.yaml:199]: Get it. I will add a doc for these new type extensions.

File: source/extensions/extensions_metadata.yaml
Comment (wbpcode) [reply source/extensions/extensions_metadata.yaml:1350]: get it.

File: source/extensions/load_balancing_policies/round_robin/config.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/round_robin/config.cc:24]: We will load and validate the configuration in the main thread. And this function will be called at runtime at worker threads.  So, IMO, a release assert would be better here.

Comment (mattklein123) [reply source/extensions/load_balancing_policies/round_robin/config.cc:24]: OK then I would make this a normal ASSERT and potentially just have a comment about why it can't be hit. Same elsewhere.

/wait

**Comment by alyssawilk** [comment]
Awesome!  Please cc me on those as I have an annual resolution to avoid more forceRegister problems for E-M and we'll probably need to add some forceRegister there :-P
cc @Augustyniak

**Review by mattklein123 — APPROVED** [review]
LGTM, thanks!

**Comment by wbpcode** [comment]
> Awesome! Please cc me on those as I have an annual resolution to avoid more forceRegister problems for E-M and we'll probably need to add some forceRegister there :-P cc @Augustyniak

cc @alyssawilk I think I have added some `forceRegister` in this PR. Could you do a quick check for it? Should I do any change?

**Comment by alyssawilk** [comment]
Sorry I'm theoretically out today so I'm going to ask Rafal to take a look.
Generally when things Envoy uses are moved to extensions we need a PR like this
https://github.com/envoyproxy/envoy/pull/24750/files
to make sure they're not dead code stripped by iOS.  I know the DFP uses its own custom host selection, but I don't recall offhand what Lyft uses for the stats cluster so I'll defer to @Augustyniak to make sure we're not losing anything Lyft uses

**Comment by Augustyniak** [comment]
we use round robin so most probably we will need to register factory related to it (in a manner similar to the PR shared by Alyssa https://github.com/envoyproxy/envoy/pull/24750/files). Trying to verify it now in https://github.com/envoyproxy/envoy/pull/24797 and will report back.

**Comment by Augustyniak** [comment]
@wbpcode mobile ci jobs in https://github.com/envoyproxy/envoy/pull/24797 are green so your changes seems to be working on mobile as-is just fine. I think that something imports/includes round robin in EM already.

In other words, no need to add any explicit factory registration for now.

**Comment by wbpcode** [comment]
> Sorry I'm theoretically out today so I'm going to ask Rafal to take a look. Generally when things Envoy uses are moved to extensions we need a PR like this https://github.com/envoyproxy/envoy/pull/24750/files to make sure they're not dead code stripped by iOS. I know the DFP uses its own custom host selection, but I don't recall offhand what Lyft uses for the stats cluster so I'll defer to @Augustyniak to make sure we're not losing anything Lyft uses

I see. This PR doesn't move the implementation of LBs to `source/extensions` but just make them could be created by the factory.
So, there should no any change to mobile for now.

**Pull request merged**
